_posts/2019-02-06-Age-Targeting-Part2.md

```diff
@@ -0,0 +1,237 @@
+We want know if the model we train on and so also the ground truth are reliable.
+We use the weights of the Logistic Regression to see which pattern can be learned from the provided features.
+
+```python
+logistic_regression = LogisticRegression(
+    maxIter=max_iter,
+    featuresCol=outputcolname,
+    labelCol=target_col,
+    elasticNetParam=elastic_net_param,
+    regParam=reg_param
+)
+pipeline_train = Pipeline(stages=[preprocessing_pipeline, logistic_regression])
+```
+
+The choice of the hyperparameters of the logistic regression and the preprocessing pipeline is preformed via cross validation.
+We use 5 folds and use the f1 metric to select the best result.
+
+```python
+params = {
+    "min_dfs_site_bundle": features_parameters["apps"],
+    "reg_params": reg_params,
+    "elastic_net_params": elastic_net_params,
+    "max_iters": max_iters
+}
+all_stages = pipeline_model.getStages()
+cv_sb = all_stages[index_for_the_apps]
+lr_model = all_stages[-1]
+grid_builder = ParamGridBuilder()
+grid_builder = grid_builder.addGrid(cv_sb.minDF, params["min_dfs_site_bundle"])
+grid_builder = grid_builder.addGrid(lr_model.elasticNetParam, params["elastic_net_params"])
+grid_builder = grid_builder.addGrid(lr_model.regParam, params["reg_params"])
+grid_builder = grid_builder.addGrid(lr_model.maxIter, params["max_iter"])
+
+evaluator = MulticlassClassificationEvaluator(labelCol=LABEL_COL, metricName="f1")
+cv = CrossValidator(
+    estimator=estimator,
+    estimatorParamMaps=grid_builder.build(),
+    evaluator=evaluator,
+    numFolds=5
+)
+cv_model = cv.fit(df)
+
+# Example to get the best parameters values:
+logging.info(
+    "regularisation parameter from best model: %f",
+    cv_model.bestModel.stages[-1]._java_obj.getRegParam() # work-around for Spark bug
+)
+best_model = cv_model.bestModel
+```
+
+It is worth mentioning that the usage and the tuning of the regularization parameters turned out to be the most crucial one for both performance and stability of the model.
+
+The weights of the logistic regression model can be interpreted as a feature importance and the top ones are shown in the picture below.
+Thus we can observe the most relevant aspects of our data which allow the model to discover the age behind a certain device.
+<p align="center"> <img src="../images/Age-Targeting/weight_class.png"> </p>
+As you can see the best features of each class are the apps used.
+Moreover, this list make extremely sense with what you would expect.
+PUT SOMETHING HERE WHICH DESCRIBE THE APPS IN THE LIST
+This gives additional confidence on putting the model into production.
+
+Analyzing the output where the model fails is often useful.
+In the plot you can see the distribution of the difference between the real age and the estimated age_group.
+Keep in mind that every class span about 10 years and so a value of the showed difference aroung -10, 10 correspond with a deviceid whose age_group is missclassified with a adjacent class.
```

Comment:
span**s**

Reply:
10 -> ten

Reply:
years. Therefore, the range from -10 to 10 corresponds to adjacent misclassified groups.